python/ray/util/spark/start_ray_node.py

```diff
@@ -8,31 +8,185 @@
 import socket
 import logging
 import threading
+import psutil
+import importlib
+import errno
+from ray.util.annotations import DeveloperAPI, PublicAPI
 
 from ray.util.spark.cluster_init import (
+    RAY_ON_SPARK_LOGGER_LEVEL,
     RAY_ON_SPARK_COLLECT_LOG_TO_PATH,
     RAY_ON_SPARK_START_RAY_PARENT_PID,
 )
 from ray._private.ray_process_reaper import SIGTERM_GRACE_PERIOD_SECONDS
-
+from ray._private import ray_constants
 
 # Spark on ray implementation does not directly invoke `ray start ...` script to create
-# ray node subprocess, instead, it creates a subprocess to run this
+# ray node subprocess, instead, it creates a child process to run this
 # `ray.util.spark.start_ray_node` module, and in this module it invokes `ray start ...`
 # script to start ray node, the purpose of `start_ray_node` module is to set up a
 # exit handler for cleaning ray temp directory when ray node exits.
 # When spark driver python process dies, or spark python worker dies, because
 # `start_ray_node` starts a daemon thread of `check_parent_alive`, it will detect
 # parent process died event and then trigger cleanup work.
 
+_logger = logging.getLogger("ray.util.spark")
+_logger.setLevel(ray_constants.LOGGER_LEVEL.upper())
 
-_logger = logging.getLogger(__name__)
+# checks for closed fds
+def is_fd_closed(lock_fd):
+    try:
+        fd_status = fcntl.fcntl(lock_fd, fcntl.F_GETFD)
+        if fd_status == 1:
+            return False
+        else:
+            return True
+    except OSError as e:
+        # Bad file descriptor error
+        if e.errno == 9: 
+            return True
+        else:
+            # Re-raise the exception if it's not a bad file descriptor error
+            raise
 
+# List process open FDs and their targets
+def list_fds():
+    if not sys.platform.startswith('linux'):
+        raise NotImplementedError('Unsupported platform: %s' % sys.platform)
+    ret = {}
+    base = '/proc/self/fd'
+    for num in os.listdir(base):
+        path = None
+        try:
+            path = os.readlink(os.path.join(base, num))
+        except OSError as err:
+            # Last FD is always the "listdir" one (which may be closed)
+            if err.errno != errno.ENOENT:
+                raise
+        ret[int(num)] = path
+    return ret
 
-if __name__ == "__main__":
+def try_clean_temp_dir_at_exit(process, lock_fd, collect_log_to_path, temp_dir):
```

Comment:
I might lose some context. Do we need to modify so many lines in `python/ray/util/spark/start_ray_node.py` file ?

if I understand it correctly, to support PEX, we just add a config to allow changing python executable path when starting ray-on-spark child process.

Reply:
Fair question(s). I modified start_ray_node.py for a number of reasons: 

1) Direct calling of ```ray.scripts.scripts.main()``` per your feedback in my previous P/R ( https://github.com/ray-project/ray/pull/44411#pullrequestreview-1991566753 ). Based on that requirement and (from my understanding of) general python community consensus regarding when/why to utilize subprocess this is possible and most efficiently done by forking the existing runtime initialized in cluster_init.py, without having the overhead of (os.fork +os.execvp) -- path scrubbing then reapplying -- and stdin/stdout redirection. I do understand the convenience of subprocess.popen from a code perspective but it's bulky when all we really needed was a light clone of the existing runtime. I specifically saw some overhead/slowness with the existing implementation (comparatively) with higher levels of logging and heavier/complex model implementations. 

2) For line 69 specifically I noticed the "try_clean_temp_dir_at_exit()" function has local constants when it could simply accept them as parameters, so in staying with the overall ray codebase designs and python community guidelines I parameterized the constants so that the function itself would be more clearly/independently defined. 

3) For code clarity, and to stay aligned with the rest of the ray codebase I created a "main()" function for the start_ray_node module, which is making it slightly look like I changed a little bit more of the codebase than I actually did. All of the "how it works architecture/functionality is the same". 

4) I changed line 46 to implement ray on spark logging as I have seen it implemented elsewhere in the ray codebase via ```ray_constants.LOGGER_LEVEL``` 

5) I added logger.debug level information for users, I think this is especially important for non-databricks users who may be implementing their own spark clusters with their own underlying filesystem types (like they might likely do on kubernetes). One place in particular this can be problematic is with file locks, I added some logging that will explain needing to change the ray temp dir/log dir or fs in this type of a situation.  

6) is_fd_closed() and list_fds() are used as more of a safety net than hard requirement type of way, that help mitigate any possible hanging file descriptors in the case of shutting down a ray cluster in an existing python/spark/notebook context and then starting up a new ray cluster within the same context. This is more of an edge case, but one that I did not want to overlook. 

7) In try_clean_temp_dir_at_exit() changes reflect using fork instead of subprocess, and more explicitly handling Exceptions, with some pretty clear debug logging. 

I'll add another comment showing a side-by-side with process logging, and some of my unit tests killing spark, calling ray shutdowns() etc while monitoring sys processes and fd(s).

Reply:
emm, this is a big change, my concern is it might introduce new bugs that unit tests can't cover. Previously I have fixed several bugs and current code has been fully tested.

can we split the work into
(1) setting python bin environmental variables for switching virtual python env 

and (2) refactor `start_ray_node.py`, avoid launching child process.

(2) is a big change, we need fully testing, do you have databricks environment ? Before merging (2) we need to fully test the change in databricks otherwise it might break our customer's workloads. Personally I don't like to do such change except we have critical bug or serious performance issue.